lib/rules/no-invalid-default-props.js

```diff
@@ -0,0 +1,624 @@
+      // If this defaultProps is "unresolved" or the propTypes is undefined, then we should ignore
+      // this component and not report any errors for it, to avoid false-positives with e.g.
+      // external defaultProps/propTypes declarations or spread operators.
+      if (defaultProps === 'unresolved' || !propTypes) {
+        return;
+      }
+
+      defaultProps.forEach(function(defaultProp) {
+        var prop = propFromName(propTypes, defaultProp.name);
+
+        if (prop && (allowRequiredDefaults || !prop.isRequired)) {
+          return;
+        }
+
+        if (prop) {
+          context.report(
+            defaultProp.node,
+            'defaultProp "{{name}}" defined for isRequired propType.',
+            {name: defaultProp.name}
+          );
+        } else {
+          context.report(
+            defaultProp.node,
+            'defaultProp "{{name}}" has no corresponding propTypes declaration.',
+            {name: defaultProp.name}
+          );
+        }
+      });
+    }
+
+    // --------------------------------------------------------------------------
+    // Public API
+    // --------------------------------------------------------------------------
+
+    return {
+      MemberExpression: function(node) {
+        var isPropType = isPropTypesDeclaration(node);
+        var isDefaultProp = isDefaultPropsDeclaration(node);
+
+        if (!isPropType && !isDefaultProp) {
+          return;
+        }
+
+        // find component this propTypes/defaultProps belongs to
+        var component = utils.getRelatedComponent(node);
+        if (!component) {
+          return;
+        }
+
+        // e.g.:
+        // MyComponent.propTypes = {
+        //   foo: React.PropTypes.string.isRequired,
+        //   bar: React.PropTypes.string
+        // };
+        //
+        // or:
+        //
+        // MyComponent.propTypes = myPropTypes;
+        if (node.parent.type === 'AssignmentExpression') {
+
+          var expression = resolveNodeValue(node.parent.right);
+          if (!expression || expression.type !== 'ObjectExpression') {
+            // If a value can't be found, we mark the defaultProps declaration as "unresolved", because
+            // we should ignore this component and not report any errors for it, to avoid false-positives
+            // with e.g. external defaultProps declarations.
+            if (isDefaultProp) {
+              markDefaultPropsAsUnresolved(component);
+            }
+
+            return;
+          }
+
+          if (isPropType) {
+            addPropTypesToComponent(component, getPropTypesFromObjectExpression(expression));
+          } else {
+            addDefaultPropsToComponent(component, getDefaultPropsFromObjectExpression(expression));
+          }
+
+          return;
+        }
+
+        // e.g.:
+        // MyComponent.propTypes.baz = React.PropTypes.string;
+        if (node.parent.type === 'MemberExpression' && node.parent.parent.type === 'AssignmentExpression') {
```

Comment:
I don't know much of anything about AST nodes, I just want to avoid a crash in the future when a new node type comes out